pkg/image/imagecontentsources/alternativesources.go

```diff
@@ -0,0 +1,134 @@
+	}
+	if len(icspData) == 0 {
+		return fmt.Errorf("no data found in ImageContentSourceFile %s", c.ICSPFile)
+	}
+	icspObj, err := runtime.Decode(operatorv1alpha1scheme.Codecs.UniversalDeserializer(), icspData)
+	if err != nil {
+		return fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", c.ICSPFile, err)
+	}
+	icsp, ok := icspObj.(*operatorv1alpha1.ImageContentSourcePolicy)
+	if !ok {
+		return fmt.Errorf("could not decode ImageContentSourcePolicy from %s", c.ICSPFile)
+	}
+	c.ICSPList = append(c.ICSPList, *icsp)
+	return nil
+}
+
+// alternativeImageSources returns list of DockerImageReference objects from list of ImageContentSourcePolicy objects
+func (a *ImageSources) alternativeImageSources(imageRef imagereference.DockerImageReference) ([]imagereference.DockerImageReference, error) {
+	var imageSources []imagereference.DockerImageReference
+	for _, icsp := range a.ICSPList {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		var sourceMatches bool
+		for _, rdm := range repoDigestMirrors {
+			rdmRef, err := imagereference.Parse(rdm.Source)
+			if err != nil {
+				return nil, err
+			}
+			if imageRef.AsRepository() == rdmRef.AsRepository() {
+				klog.V(2).Infof("%v RepositoryDigestMirrors source matches given image", imageRef.AsRepository())
+				sourceMatches = true
+			}
+			for _, m := range rdm.Mirrors {
+				if sourceMatches {
+					klog.V(2).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
+					mRef, err := imagereference.Parse(m)
+					if err != nil {
+						return nil, err
+					}
+					imageSources = append(imageSources, mRef)
+				}
+			}
+		}
+	}
+	uniqueMirrors := make([]imagereference.DockerImageReference, 0, len(imageSources))
+	uniqueMap := make(map[imagereference.DockerImageReference]bool)
+	for _, imageSourceMirror := range imageSources {
+		if _, ok := uniqueMap[imageSourceMirror]; !ok {
+			uniqueMap[imageSourceMirror] = true
+			uniqueMirrors = append(uniqueMirrors, imageSourceMirror)
+		}
+	}
+	// make sure at least 1 imagesource
+	// ie, make sure the image passed is included in image sources
+	// this is so the user-given image ref will be tried
+	if len(imageSources) == 0 {
+		imageSources = append(imageSources, imageRef.AsRepository())
+		return imageSources, nil
+	}
+	klog.V(2).Infof("Found sources: %v for image: %v", uniqueMirrors, imageRef)
+	return uniqueMirrors, nil
+}
+
+// Alternative returns possible DockerImageReferences
+// from image sources gathered from ImageContentSourcePolicy objects and user-passed image.
+// Alternative will lookup from cluster or from ImageContentSourcePolicy file passed from user.
+// Image reference of user-given image may be different from original in case of mirrored images.
+// Satisfies registrycontext.ImageSources interface
+func (a *ImageSources) Alternative(image imagereference.DockerImageReference) ([]imagereference.DockerImageReference, error) {
```

Comment:
all that's required by Repository call is the registry URL and the repoName (string), so I can rework alternative fns to only return those, rather than the []DockerImageReference